add gcp kubeadm clusterclass example (#327)
rancher/turtles-docs: docs/next/modules/en/pages/user/clusterclass.adoc

      matchLabels: {}
----

GCP::
+

> To prepare the management Cluster, we are going to install the https://cluster-api-gcp.sigs.k8s.io/[Cluster API Provider GCP], and create a secret with the required credentials to provision a new Cluster on GCP. +

Reviewer: nit:
```suggestion
To prepare the management Cluster, we are going to install the https://cluster-api-gcp.sigs.k8s.io/[Cluster API Provider GCP], and create a secret with the credentials required to provision a new Cluster on GCP. +
```

A Service Account is required to create and manage clusters in GCP and this will require `Editor` permissions. You can follow the offical guide from the https://cluster-api-gcp.sigs.k8s.io/quick-start#create-a-service-account[CAPG Book]. +
+
* Provider installation
+
[source,yaml]
----
apiVersion: turtles-capi.cattle.io/v1alpha1
kind: CAPIProvider
metadata:
  name: gcp
  namespace: capg-system
spec:
  type: infrastructure
  variables:

>     GCP_B64ENCODED_CREDENTIALS: ""

Reviewer: This part is a little unclear to me. Do I pass the encoded value to this variable or leave it empty?

====
+
* A GCP Kubeadm ClusterClass can be found among the https://github.com/rancher/turtles/tree/main/examples/clusterclasses[Turtles examples].
+
Applications like https://docs.tigera.io/calico/latest/about/[Calico CNI] and https://github.com/kubernetes/cloud-provider-gcp[Cloud Controller Manager GCP] will be installed on downstream Clusters. This is done automatically at Cluster creation by targeted Clusters with specific labels, such as `cni: calico` and `cloud-provider: gcp`.
+
[tabs]
=======
CLI::
+
A GCP Kubeadm ClusterClass and associated applications can be applied using the examples tool:
+
[source,bash]
----
go run github.com/rancher/turtles/examples@latest -r gcp-kubeadm | kubectl apply -f -
----

kubectl::
+
* Alternatively, you can apply the GCP Kubeadm ClusterClass directly using kubectl:
+
[source,bash]
----
kubectl apply -f https://raw.githubusercontent.com/rancher/turtles/refs/heads/main/examples/clusterclasses/gcp/kubeadm/clusterclass-kubeadm-example.yaml
----
+
* For this example we are also going to install https://docs.tigera.io/calico/latest/about/[Calico] as the default CNI. +

> * The https://github.com/kubernetes/cloud-provider-gcp[Cloud Controller Manager GCP] will need to be installed on each downstream Cluster for the nodes to be functional. +

Reviewer: nit:
Can we call it `GCP Cloud Controller Manager` instead of `Cloud Controller Manager GCP`? It seems easier to me to read and contemplate.

Author: I agree your suggestion sounds more natural. This is taken from the naming used in AWS, which is also `Cloud Controller Manager AWS`. If we change this one, it'll make sense to have other occurrences changed, too.

Reviewer: I am okay with changing AWS, but if you'd like to do it in a separate PR that is fine too.

+
We can do this automatically at Cluster creation using a combination of https://rancher.github.io/cluster-api-addon-provider-fleet/[Cluster API Add-on Provider Fleet] and https://fleet.rancher.io/bundle-add[Fleet Bundle]. +
The Add-on provider is installed by default with {product_name}. +
The `HelmApps` need to be created first, to be applied on the new Cluster via label selectors. This will take care of deploying Calico. +
+
[source,bash]
----

> kubectl apply -f https://raw.githubusercontent.com/rancher/turtles/refs/heads/main/examples/applications/cni/aws/calico/helm-chart.yaml

Author: Thanks for catching this! Copy/paste failure.

----
+
A `Bundle` will take care of deploying Cloud Controller Manager GCP. The reason for not using Add-on Provider Fleet is that https://github.com/kubernetes/cloud-provider-gcp[Cloud Controller Manager GCP] does not provide a Helm chart, so we opt for creating the Fleet `Bundle` resource directly. +
+
[source,bash]
----
kubectl apply -f https://raw.githubusercontent.com/rancher/turtles/refs/heads/main/examples/applications/ccm/gcp/bundle.yaml
----
=======
+
* Create the GCP Cluster from the example ClusterClass +
+ 
Note that some variables are left to the user to substitute. +

> The default configuration of Cloud Controller Manager GCP is configured to use a single zone cluster, so the `clusterFailureDomains` variable is set to a single zone. If you need to provision a multi-zone cluster, we recommend you inspect the parameters provided by https://github.com/kubernetes/cloud-provider-gcp/blob/12f93cb23a5af58bfb7fb453bebff3eb2f81755c/providers/gce/gce.go#L120[Cloud Controller Manager GCP] and how https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/6518ef9b44cfc4f8c3f7139b2ce4ae71523deff6/test/e2e/data/infrastructure-gcp/cluster-template-ci.yaml#L59[CAPG leverages these variables] to create cluster-specific configurations. +

Reviewer: nit:
```suggestion
The default configuration of Cloud Controller Manager GCP is configured to use a single zone cluster, so the `clusterFailureDomains` variable is set to a single zone. If you need to provision a multi-zone cluster, we recommend you inspect the parameters provided by https://github.com/kubernetes/cloud-provider-gcp/blob/master/providers/gce/gce.go#L120[Cloud Controller Manager GCP] and how https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/main/test/e2e/data/infrastructure-gcp/cluster-template-ci.yaml#L59[CAPG leverages these variables] to create cluster-specific configurations. +
```

+
[source,yaml]
----

apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  labels:
    cluster-api.cattle.io/rancher-auto-import: "true"
    cni: calico
    cloud-provider: gcp
  name: gcp-quickstart
spec:
  clusterNetwork:
    pods:
      cidrBlocks: 
      - 192.168.0.0/16
  topology:
    class: gcp-kubeadm-example
    controlPlane:
      replicas: 1
    workers:
      machineDeployments:
        - class: "default-worker"
          name: "md-0"
          replicas: 1
    variables:
      - name: gcpProject
        value: <GCP_PROJECT>
      - name: region
        value: <GCP_REGION>
      - name: gcpNetworkName
        value: <GCP_NETWORK_NAME>
      - name: clusterFailureDomains
        value:
          - "<GCP_REGION>-a"
      - name: imageId
        value: <GCP_IMAGE_ID>
      - name: machineType
        value: <GCP_MACHINE_TYPE>

>     version: v1.31.7

Reviewer 2: ```suggestion
    version: v1.31.4
```

This is the e2e validated version.
